modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java

```diff
@@ -5524,58 +5527,91 @@ private void restorePartitionStates(
+
+                    SortedSet<T3<Long, Long, GroupPartitionId>> top =
+                        new ConcurrentSkipListSet<>(processedPartitionComparator());
+
+                    while (!batch.isEmpty()) {
+                        GroupPartitionId grpPartId = batch.poll();
+
+                        CacheGroupContext grpCtx = ctx.cache().cacheGroup(grpPartId.getGroupId());
+
+                        try {
+                            long time = grpCtx.offheap().restoreStateOfPartition(grpPartId.getPartitionId(),
+                                partStates.get(grpPartId));
 
-                                for (T3<Long, Long, GroupPartitionId> t2 : top0) {
-                                    top.add(t2);
+                            top.add(new T3<>(time, U.currentTimeMillis(), grpPartId));
 
-                                    trimToSize(top, 5);
-                                }
+                            trimToSize(top, topPartRefLimit);
 
-                                return top;
-                            });
+                            totalProcessed.incrementAndGet();
                         }
-                    }
-                    catch (IgniteCheckedException | RuntimeException | Error e) {
-                        U.error(log, "Failed to restore partition state for " +
-                            "groupName=" + grp.name() + " groupId=" + grp.groupId(), e);
+                        catch (IgniteCheckedException | RuntimeException | Error e) {
+                            U.error(log, "Failed to restore partition state for " +
+                                "groupName=" + grpCtx.name() + " groupId=" + grpCtx.groupId(), e);
 
-                        restoreStateError.compareAndSet(
-                            null,
-                            e instanceof IgniteCheckedException
+                            IgniteCheckedException ex = e instanceof IgniteCheckedException
                                 ? ((IgniteCheckedException)e)
-                                : new IgniteCheckedException(e)
-                        );
-                    }
-                    finally {
-                        completionLatch.countDown();
+                                : new IgniteCheckedException(e);
+
+                            if (!restoreStateError.compareAndSet(null, ex))
+                                restoreStateError.get().addSuppressed(ex);
+                        }
+                        finally {
+                            completionLatch.countDown();
+
+                            AtomicInteger completedCntr = grps.get(grpPartId.getGroupId());
+
+                            if (completedCntr.incrementAndGet() == grpCtx.affinity().partitions())
+                                grpCtx.offheap().confirmPartitionStatesRestored();
+
+                            if (log.isInfoEnabled()) {
+                                topPartRef.updateAndGet(top0 -> {
```

Comment:
This code is executed for each partition, then the local collection (`SortedSet<T3<Long, Long, GroupPartitionId>> top`) is not needed, I think this will be a place of high contention, perhaps you need to think about how to reduce / remove it.